### ISSUE TITLE
Issue with hypermorph (--image-sigma and sampling distribution)

### ISSUE THREAD

**Issue by xiaogengen-007** [voxelmorph/voxelmorph#419]
**Task** (what are you trying to do/register?)

Hello, I am trying to register 3D head&neck CT images and use hypermorph to find the best regularization parameters.

**What have you tried**

Question 1 (--image-sigma):
I first trained with the default --image-sigma in train_hypermorph.py (0.05). The results didn't look good probably due to the mse loss being scaled too large. Then I tried --image-sigma =1 and it worked much better. I found in another issue saying --image-sigma hasn't been really used (https://github.com/voxelmorph/voxelmorph/issues/408). So I just want to make sure: do I need to have a good estimation of image noise or I can just ignore this hyperparameter by setting it to 1.

Question 2 (hyperparameter sampling distribution):
L = L_sim + γ * L_reg (γ ∈ [0,+∞])
L = (1-λ) * L_sim + λ * L_reg (λ ∈ [0,1])
I noticed in train_hypermorph.py that λ is uniformly sampled from [0,1]. If the optimal γ is too small or too large (e.g. γ* = 0.01, λ* = 0.0099; γ* = 100, λ* = 0.99), very few samples will be drawn around the optimal value. Will this be an issue? My personal intuition is that if we don't have any prior knowledge about the optimal λ or γ, we could design the distribution of λ such that the corresponding distribution of γ is log uniform. The end-point oversampling (code below) seems to be addressing this issue, but why only oversampling exactly at the endpoint instead of oversampling around the endpoint. I am wondering if you have already conducted any experiment showing that this strategy is robust no matter what value the optimal λ takes?                                                                                                                                                                                                                                                                                                                                                    
```
# random hyperparameter generator
def random_hyperparam():
    if np.random.rand() < args.oversample_rate:
        return np.random.choice([0, 1])
    else:
        return np.random.rand()              
```                                                                                                                                                                                                           

Thanks in advance for any help and thank you for this fantastic work!

**Comment by adalca** [comment]
I'll let @ahoopes comment in more detail, but in the journal version of HyperMorph (https://arxiv.org/abs/2203.16680) we discuss the importance of normalizing your loss terms (Figure 12). One good thing to do is normalize your data (or the MSE term) by the noise in your data.  That's why we ask for the image noise. If you estimate it in the right ballpark, you'll get a better behaved overall hyperparameter landscape (blue in Fig 12). If it's a poor estimate, you can get a very badly behaved curve (black in Fig 12).

**Comment by ahoopes** [comment]
Right, as Adrian mentioned, the sigma term is very important for normalization in MSE. You could try a value of 1, which might work fine in some cases, but you risk skewing the landscape of reasonable lambda values to a very small range.

At the moment, I'm updating some hypermorph test scripts so that we can better visualize these effects.

**Comment by xiaogengen-007** [comment]
l

**Comment by xiaogengen-007** [comment]
Thank you both for the great comments!!!
I tested a few values for --image-noise (0.05, 0.2, and 1) and got something similar to Figure 12 in the paper. It seems sigma = 0.2 can give me the best-behaved curve for lambda.
![hypermorph_image_noise_sigma](https://user-images.githubusercontent.com/45641353/170214997-4ab36c80-1caf-4491-9391-a6f12753540b.png)
My further question is does this sigma value of 0.2 rigorously correspond to the image noise of the data? 
I am using CT images and normalized from -1024 ~ +3000 HU to 0~1. So image noise of 0.2 would correspond to ~ 800 HU, which doesn't make sense to me.

I am not sure what is wrong but the following are my random thoughts:
> I am thinking of this sigma as a scaling constant instead of the real image noise. I guess for MRI image reconstruction, it is fair to normalize the MSE by image noise as we know the noise is Gaussian. But in image registration, the MSE seems to be a mixture of Gaussian noise and registration error (non-Gaussian). So, I am wondering in this case, does this optimal sigma value still have the physical meaning of image noise.

Thank you in advance for the help!

**Comment by xiaogengen-007** [comment]
Thank you both for the great comments!!!
I tested a few values for --image-noise (0.05, 0.2, and 1) and got something similar to Figure 12 in the paper. It seems sigma = 0.2 can give me the best-behaved curve for lambda.
![hypermorph_image_noise_sigma](https://user-images.githubusercontent.com/45641353/170223076-b8a81286-f1bd-4398-8b89-a11329278900.png)
My further question is does this sigma value of 0.2 rigorously correspond to the image noise of the data?
I am using CT images and normalized from -1024 ~ +3000 HU to 0~1. So image noise of 0.2 would correspond to ~ 800 HU, which doesn't make sense to me.

I am not sure what is wrong but the following are my random thoughts:
> I am thinking of this sigma as a scaling constant instead of the real image noise. I guess for MRI image reconstruction, it is fair to normalize the MSE by image noise as we know the noise is Gaussian. But in image registration, the MSE seems to be a mixture of Gaussian noise and registration error (non-Gaussian). So, I am wondering in this case, does this optimal sigma value still have the physical meaning of image noise.

Thank you in advance for any help!

**Comment by ahoopes** [comment]
In reality, sigma shouldn't really represent intrinsic 'image noise' across a dataset. It's actually the image noise *between* two images. It could be computed as mean square difference between images assuming a 'perfect' anatomical alignment.

So, this value would likely exceed the intrinsic image noise because it's the compounded noise between two images and inherently non-alignable regions would have substantial intensity differences.

That being said, I can't say for sure how rigorously sigma (or the 'ideal' sigma) corresponds to noise, but it's something that would be worth experimenting with in detail. Though we expect that sigma representing this noise is a pretty good model / estimation. Hope that helps!

**Comment by xiaogengen-007** [comment]
> In reality, sigma shouldn't really represent intrinsic 'image noise' across a dataset. It's actually the image noise _between_ two images. It could be computed as mean square difference between images assuming a 'perfect' anatomical alignment.
> 
> So, this value would likely exceed the intrinsic image noise because it's the compounded noise between two images and inherently non-alignable regions would have substantial intensity differences.
> 
> That being said, I can't say for sure how rigorously sigma (or the 'ideal' sigma) corresponds to noise, but it's something that would be worth experimenting with in detail. Though we expect that sigma representing this noise is a pretty good model / estimation. Hope that helps!

Thank you for the clear explanation! That makes sense to me. 
So, if I am understanding it correctly, this sigma needs to be tuned when working on a new dataset. Or at least we need to run hypermorph with an estimated sigma and then we can calculate a better estimation of it based on the Dice-lambda curve. Is that correct, or does a better way to estimate this sigma exist?

**Comment by adalca** [comment]
Well hopefully you can do a back-of-the-envelope estimation of sigma 
(estimate it as higher than your noise) and it gets you in the right 
ballpark, and from there on you can just use hypermorph to plot the 
whole spectrum for lambda.
